internal/clients/team4/president.go

```diff
@@ -0,0 +1,351 @@
+
+	if budgetRuleInPay && actionOverBudget {
+		return shared.PresidentReturnContent{
+			ContentType: shared.PresidentRuleProposal,
+			ActionTaken: false,
+		}
+	}
+
+	// DefaultContentType: No rules were proposed by the islands
+	proposedRuleMatrix := rules.RuleMatrix{}
+	actionTaken := false
+
+	// if some rules were proposed
+	//TODO: Pick rules close to ideals
+	if len(rulesProposals) != 0 {
+		proposedRuleMatrix = rulesProposals[rand.Intn(len(rulesProposals))]
+		actionTaken = true
+	}
+
+	return shared.PresidentReturnContent{
+		ContentType:        shared.PresidentRuleProposal,
+		ProposedRuleMatrix: proposedRuleMatrix,
+		ActionTaken:        actionTaken,
+	}
+}
+
+// SetTaxationAmount sets taxation amount for all of the living islands
+// islandsResources: map of all the living islands and their reported resources
+func (p *president) SetTaxationAmount(islandsResources map[shared.ClientID]shared.ResourcesReport) shared.PresidentReturnContent {
+	_, budgetRuleInPay := p.GameState.RulesInfo.CurrentRulesInPlay["president_over_budget"]
+	actionOverBudget := p.parent.ServerReadHandle.GetGameState().IIGORolesBudget[shared.President]-
+		p.parent.ServerReadHandle.GetGameConfig().IIGOClientConfig.BroadcastTaxationActionCost < 0
+
+	if budgetRuleInPay && actionOverBudget {
+		return shared.PresidentReturnContent{
+			ContentType: shared.PresidentTaxation,
+			ActionTaken: false,
+		}
+	}
+
+	taxAmountMap := make(map[shared.ClientID]shared.Resources)
+
+	for clientID, clientReport := range islandsResources {
+		weAreRichThreshold, weAreRichThresholdIsApplicable := p.generateWeAreRichThreshold()
+		if weAreRichThresholdIsApplicable && p.parent.ServerReadHandle.GetGameState().CommonPool > weAreRichThreshold {
+			//Excuse everyone if its not season one and there are a lot of resources
+			taxAmountMap[clientID] = shared.Resources(0)
+		} else if p.parent.ServerReadHandle.GetGameState().ClientLifeStatuses[clientID] == shared.Critical {
+			//Excuse if the client is critical
+			taxAmountMap[clientID] = shared.Resources(0)
+		} else if clientReport.Reported {
+			//Excuse if the reports have to be true and the island is poor
+			//TODO: remove magik numbers
+			_, reportTruthfulnessRuleInPay := p.GameState.RulesInfo.CurrentRulesInPlay["island_must_report_actual_private_resource"]
+			if reportTruthfulnessRuleInPay && clientReport.ReportedAmount*3 <= p.parent.ServerReadHandle.GetGameConfig().CostOfLiving {
+				taxAmountMap[clientID] = shared.Resources(0)
+			} else if clientReport.ReportedAmount*1.5 <= p.parent.ServerReadHandle.GetGameConfig().CostOfLiving {
+				taxAmountMap[clientID] = shared.Resources(0)
+			}
+			taxAmountMap[clientID] = p.generateTaxAmount(clientID, clientReport.ReportedAmount)
+		} else {
+			taxAmountMap[clientID] = 15 //flat tax rate
+		}
+
+		//Excuse ourselves if we are not fair
+		if clientID == shared.Team4 {
+			taxAmountMap[clientID] = shared.Resources(float64(taxAmountMap[clientID]) * p.parent.internalParam.fairness)
```

Comment:
How about having a threshold on this, so it has to be over for example 0.2 before we start modifying the tax